scriptmodules/supplementary/skyscraper.sh

```diff
@@ -287,10 +300,12 @@ function _scrape_skyscraper() {
         params+=(--refresh)
     fi
 
+    local param_list="${params[@]}"
```

Comment:
Why the additional variable rather than using the array directly?

Reply:
I saw on the forum it's due to variable expansion, but shouldn't need a new var. Did you try `${params[*]}`

Reply:
I didn't, but I'll give it a shot and see if I can improve it.

Reply:
It turns out it was the quoting was passing the whole params array as a parameter. Scraper does this correctly [here](https://github.com/RetroPie/RetroPie-Setup/blob/9549a01a44308af922daa7c8acd8a4269dfdd0b2/scriptmodules/supplementary/scraper.sh#L136), I've must have added it myself and broke it during the review.
I've added another fix found during testing for this modification.

Reply:
Interesting. To my understanding the quotes should actually be there to prevent word splitting. With quotes the paramters are expanded, quoted and passed one by one. See this quick test:
```sh
#!/usr/bin/env bash
function print() { local arg; for arg in "$@"; do echo "argument: $arg"; done}
myarray=(this is a "test with" five words)
print "${myarray[@]}"
```
Which works as expected:
```
$ bash test.sh
argument: this
argument: is
argument: a
argument: test with
argument: five
argument: words
```
Maybe there is something else giving you weird results. The `sudo` call perhaps?

Reply:
And if you remove the quotes:
```sh
print ${myarray[@]}
```
You get the undesired effect of word splitting:
```
$ bash test.sh
argument: this
argument: is
argument: a
argument: test
argument: with
argument: five
argument: words
```

Reply:
Interesting, I'll do a test to see how this runs through `sudo`.

Reply:
It doesn't make a difference, using your test, if using `sudo` or not. Quoted expansion happens every time.

Reply:
What is the exact error you receive when quoting the expanded array? I'm very curious now.

I just tested through `sudo` like this:
```sh
#!/usr/bin/env bash
myarray=(/boot /mnt /proc "no exist")
sudo ls -1lad "${myarray[@]}"
```
And it works as expected (notice how `ls` can't find the single dir `no exist` and does find each element of the array correctly):
```
$ bash test.sh
ls: cannot access no exist: No such file or directory
drwxr-xr-x   4 root root 2048 May  2  2015 /boot
drwxr-xr-x   4 root root 4096 Jan 10  2014 /mnt
dr-xr-xr-x 255 root root    0 May 20 14:04 /proc
```
I'm skeptic of the program mishandling the quotes because the shell expands all arguments before even calling the program.

Reply:
@hhromic there's no error quoting the array in the script, but I found what happens. The `$relative` variable can be empty and it's passed as a quoted empty string by `"${myarray[@]}"` , which does not happen for `${myarray[@]}` . `Skyscraper` gets the empty parameter and assumes it's a filename, thus the error message.

@joolswills I'll do a proper fix for this.

Reply:
brilliant! well done :) that was tricky to spot

Reply:
@joolswills fixed by selectively adding `$relative`.

Reply:
I got thinking, why not initialise `params` further up (before the conditionals) and then just add parameters according to the conditions? you could get rid of the temporary `gamelist`, `media_dir` and `relative` variables all-together like that. For example:
```sh
local -a params=("--unattend" "--skipped")
if [[ "$use_rom_folder" -eq 1 ]]; then
    params+=(-g "$romdir/$system")
    params+=(-o "$romdir/$system/media")
    # If we're saving to the ROM folder, then use relative paths in the gamelist
    params+=(--relative)
else
    params+=(-g "$home/.emulationstation/gamelists/$system")
    params+=(-o "$home/.emulationstation/downloaded_media/$system")
fi
```
I think it makes cleaner and more robust code :)

Reply:
You could also compact code like this for less clutter:
```sh
if [[ "$cache_screenshots" -eq 0 ]]; then
    params+=(--noscreenshots)
fi
# is equivalent to this
[[ "$cache_screenshots" -eq 0 ]] && params+=(--noscreenshots)
```

Reply:
@hhromic thanks for the tips! I'll probably peruse the script for possible optimizations and add your observations to it, in a subsequent revision.